vim-patch:8.1.{569,571}

janlazo opened the pull request:
closes #6035

**vim-patch:8.1.0569: execute() always resets display column to zero**

Problem:    Execute() always resets display column to zero. (Sha Liu)
Solution:   Don't reset it to zero, restore the previous value. (closes vim/vim#3669)
vim/vim@10ccaa1

**vim-patch:8.1.0571: non-silent execute() resets display column to zero**

Problem:    Non-silent execute() resets display column to zero.
Solution:   Keep the display column as-is.
vim/vim@446e7a3

liushapku committed:
UI: Fix wrong msg_col after execute()

closes #6035
closes #9250

liushapku committed:
update functional test for "places cursor correctly #6035"

janlazo committed:
vim-patch:8.1.0569: execute() always resets display column to zero

Problem:    Execute() always resets display column to zero. (Sha Liu)
Solution:   Don't reset it to zero, restore the previous value. (closes vim/vim#3669)
https://github.com/vim/vim/commit/10ccaa17ec8b2be1132fd19059e1cd5fb5c902c4

janlazo committed:
vim-patch:8.1.0571: non-silent execute() resets display column to zero

Problem:    Non-silent execute() resets display column to zero.
Solution:   Keep the display column as-is.
https://github.com/vim/vim/commit/446e7a3cd36b2de7d559f167eb5795d1e1cd3ddb

liushapku commented:
Could you please also add those tests in the previous [thread](https://github.com/neovim/neovim/pull/9319#issuecomment-445543720)?

liushapku commented:
Thanks. But could you please also add those lua tests from the previous [PR](https://github.com/neovim/neovim/pull/9319#issuecomment-445543720)

janlazo committed:
functionaltests: fix new execute() tests

janlazo commented:
```
From test_alot.vim:
Found errors in Test_execute_does_not_change_col():
function RunTheTest[35]..Test_execute_does_not_change_col line 9: Expected 'abcdxyz' but got '       '
Found errors in Test_execute_not_silent():
function RunTheTest[35]..Test_execute_not_silent line 9: Expected 'abcd234 ' but got '[No Name'
function RunTheTest[35]..Test_execute_not_silent line 14: Expected 'xyz ' but got '    '
```

justinmk commented:
Those failures are most likely a result of unexpected defaults in the Vim tests. E.g. Vim doesn't show a statusbar by default, so that's where `[No Name]` comes from.

Should be OK to `:set laststatus=1` in `testdir/setup.vim` .

janlazo committed:
oldtests: set laststatus=1

erw7 commented:
@janlazo The cause of https://github.com/neovim/neovim/pull/9338#issuecomment-457831520 seems to be that [`msg_use_printf()`](https://github.com/neovim/neovim/blob/773bdd41ec19c9909e979c4580cf098d71d168d7/src/nvim/message.c#L2301) returns `true` in headless mode. In my environment, applying the following patch will pass that test. However, other tests have not been verified, so there may be something wrong.

```diff
diff --git a/src/nvim/message.c b/src/nvim/message.c
index b4aa333a4..93c2675fd 100644
--- a/src/nvim/message.c
+++ b/src/nvim/message.c
@@ -2289,7 +2289,7 @@ static void t_puts(int *t_col, const char_u *t_s, const char_u *s, int attr)
 //    - no UI and not embedded
 int msg_use_printf(void)
 {
-  return !embedded_mode && !ui_active();
+  return !embedded_mode && !headless_mode && !ui_active();
 }

 /// Print a message when there is no valid screen.
```

erw7 commented:
I read `:h --headless`, the https://github.com/neovim/neovim/pull/9829#issuecomment-484034791  change is not desirable. Therefore, do not use headless mode with oldtest, or in the case of the headless mode, the `:echo` output needs to be changed to the screen and `stderr`.

justinmk commented:
> Therefore, do not use headless mode with oldtest, or in the case of the headless mode, the `:echo` output needs to be changed to the screen and `stderr`.

@erw7 what do you mean by "changed to the screen and stderr"? `--headless` already outputs to `stderr`.

```
$ 2>/dev/null nvim --headless +'echo "foo"' +q
$ 1>/dev/null nvim --headless +'echo "foo"' +q
foo
```

erw7 commented:
> what do you mean by "changed to the screen and stderr"?

@justinmk It is to change it like the following patch.

```diff
diff --git a/src/nvim/message.c b/src/nvim/message.c
index f7c248184..d793bcbad 100644
--- a/src/nvim/message.c
+++ b/src/nvim/message.c
@@ -1818,7 +1818,8 @@ void msg_puts_attr_len(const char *const str, const ptrdiff_t len, int attr)
   // cursor is.
   if (msg_use_printf()) {
     msg_puts_printf(str, len);
-  } else {
+  }
+  if (!msg_use_printf() || headless_mode) {
     msg_puts_display((const char_u *)str, len, attr, false);
   }
 }
```

justinmk commented:
Ah, I think I understand: in order to make `screenchar()` work, headless-mode must call `msg_puts_display()` so that the internal screen is updated.

We could try that...

erw7 committed:
Fix screenchar() problem in headless mode

In the case of the headless mode, screenchar() does not operate normally
because it is not output to the internal screen. Change output to stderr
and internal screen to fix it.

jityao committed:
Fix multiple c_CTRL-D showing statusline

jityao committed:
Add test

bfredl committed:
test: don't detach screen just to change the size

bfredl committed:
Merge pull request #10107 from gelguy/c_ctrl_d

Fix multiple <C-D> showing statusline on previous completion list

bfredl committed:
api/buffer: create new buffers in the "opened" state

Otherwise vim will think that ml_append() needs to "enter" the buffer,
which emits unexpected autocommands.

ref https://github.com/vim-airline/vim-airline/issues/1930

bfredl committed:
Merge pull request #10112 from bfredl/noautobuf

api/buffer: avoid spurios Buf[Win]Enter in nvim_buf_set_lines

bfredl committed:
msg_puts_attr_len: check default_grid.chars if headless

Nvim might call `msg_puts_attr_len` before the screen buffers are allocated.

justinmk committed:
Merge #10109 from erw7/fix-screenchar-on-headless-mode

ref #9338

justinmk committed:
Merge #9338 'vim-patch:8.1.{569,571}'

justinmk commented:
Updated after https://github.com/neovim/neovim/pull/10109 .
